File: Vanta Sppech Android/app/src/main/java/com/vanta/speech/core/eas/model/EASError.kt
package com.vanta.speech.core.eas.model

/**
 * Errors that can occur during EAS operations
 */
sealed class EASError : Exception() {

    /** No network connection available */
    data object Offline : EASError()

    /** No credentials stored */
    data object NoCredentials : EASError()

    /** Invalid or expired credentials (HTTP 401) */
    data object AuthenticationFailed : EASError()

    /** Access denied by server (HTTP 403) */
    data object AccessDenied : EASError()

    /** Server temporarily unavailable (HTTP 503) */
    data object ServerUnavailable : EASError()

    /** Calendar folder not found in FolderSync response */
    data object CalendarFolderNotFound : EASError()

    /** Failed to parse server response */
    data class ParseError(val details: String) : EASError()

    /** Network error during request */
    data class NetworkError(val details: String) : EASError()

    /** HTTP error with status code */
    data class ServerError(val statusCode: Int, val serverMessage: String?) : EASError()

    /** WBXML encoding/decoding error */
    data class WBXMLError(val details: String) : EASError()

    /** Invalid server URL */
    data object InvalidServerURL : EASError()

    /** Unknown error */
    data class Unknown(val details: String) : EASError()

    /**
     * User-friendly error description in Russian
     */
    val errorDescription: String
        get() = when (this) {
            is Offline -> "Нет подключения к сети. Проверьте интернет-соединение."
            is NoCredentials -> "Требуется вход в аккаунт"
            is AuthenticationFailed -> "Неверный логин или пароль"
            is AccessDenied -> "Доступ запрещён. Обратитесь к администратору."
            is ServerUnavailable -> "Сервер временно недоступен. Попробуйте позже."
            is CalendarFolderNotFound -> "Календарь не найден. Обратитесь к администратору."
            is ParseError -> "Ошибка обработки данных: $details"
            is NetworkError -> "Ошибка сети: $details"
            is ServerError -> if (serverMessage != null) {
                "Ошибка сервера ($statusCode): $serverMessage"
            } else {
                "Ошибка сервера: $statusCode"
            }
            is WBXMLError -> "Ошибка формата данных: $details"
            is InvalidServerURL -> "Неверный адрес сервера"
            is Unknown -> "Произошла ошибка: $details"
        }

    /**
     * Suggested action to recover from error
     */
    val recoverySuggestion: RecoveryAction
        get() = when (this) {
            is Offline, is NetworkError, is ServerUnavailable, is ParseError, is WBXMLError ->
                RecoveryAction.RETRY
            is NoCredentials, is AuthenticationFailed ->
                RecoveryAction.LOGIN
            is AccessDenied, is CalendarFolderNotFound ->
                RecoveryAction.CONTACT_ADMIN
            is ServerError -> when (statusCode) {
                401 -> RecoveryAction.LOGIN
                in 500..599 -> RecoveryAction.RETRY
                else -> RecoveryAction.CONTACT_ADMIN
            }
            is InvalidServerURL -> RecoveryAction.LOGIN
            is Unknown -> RecoveryAction.RETRY
        }

    /**
     * Whether credentials should be cleared for this error
     */
    val shouldClearCredentials: Boolean
        get() = when (this) {
            is AuthenticationFailed -> true
            is ServerError -> statusCode == 401
            else -> false
        }
}

/**
 * Suggested recovery action for errors
 */
enum class RecoveryAction {
    /** User should retry the operation */
    RETRY,

    /** User should re-authenticate */
    LOGIN,

    /** User should contact system administrator */
    CONTACT_ADMIN
}

File: Vanta Speech iOS/Vanta Speech/Core/EAS/Models/EASError.swift
import Foundation

/// Errors that can occur during EAS operations
enum EASError: LocalizedError, Equatable {
    /// No network connection available
    case offline

    /// No credentials stored
    case noCredentials

    /// Invalid or expired credentials (HTTP 401)
    case authenticationFailed

    /// Access denied by server (HTTP 403)
    case accessDenied

    /// Server temporarily unavailable (HTTP 503)
    case serverUnavailable

    /// Calendar folder not found in FolderSync response
    case calendarFolderNotFound

    /// Failed to parse server response
    case parseError(String)

    /// Network error during request
    case networkError(String)

    /// HTTP error with status code
    case serverError(statusCode: Int, message: String?)

    /// WBXML encoding/decoding error
    case wbxmlError(String)

    /// Invalid server URL
    case invalidServerURL

    /// Device provisioning denied by server
    case provisioningDenied

    /// Provisioning required (status 108)
    case provisioningRequired

    /// Unknown error
    case unknown(String)

    // MARK: - LocalizedError

    var errorDescription: String? {
        switch self {
        case .offline:
            return "Нет подключения к сети. Проверьте интернет-соединение."
        case .noCredentials:
            return "Требуется вход в аккаунт"
        case .authenticationFailed:
            return "Неверный логин или пароль"
        case .accessDenied:
            return "Доступ запрещён. Обратитесь к администратору."
        case .serverUnavailable:
            return "Сервер временно недоступен. Попробуйте позже."
        case .calendarFolderNotFound:
            return "Календарь не найден. Обратитесь к администратору."
        case .parseError(let details):
            return "Ошибка обработки данных: \(details)"
        case .networkError(let details):
            return "Ошибка сети: \(details)"
        case .serverError(let code, let message):
            if let message = message {
                return "Ошибка сервера (\(code)): \(message)"
            }
            return "Ошибка сервера: \(code)"
        case .wbxmlError(let details):
            return "Ошибка формата данных: \(details)"
        case .invalidServerURL:
            return "Неверный адрес сервера"
        case .provisioningDenied:
            return "Устройство заблокировано политиками безопасности. Обратитесь к администратору."
        case .provisioningRequired:
            return "Требуется принятие политик безопасности"
        case .unknown(let details):
            return "Произошла ошибка: \(details)"
        }
    }

    // MARK: - Recovery Suggestion

    /// Suggested action to recover from error
    var recoverySuggestion: RecoveryAction {
        switch self {
        case .offline, .networkError, .serverUnavailable, .parseError, .wbxmlError, .provisioningRequired:
            return .retry
        case .noCredentials, .authenticationFailed:
            return .login
        case .accessDenied, .calendarFolderNotFound, .provisioningDenied:
            return .contactAdmin
        case .serverError(let code, _):
            if code == 401 {
                return .login
            } else if code >= 500 {
                return .retry
            }
            return .contactAdmin
        case .invalidServerURL:
            return .login
        case .unknown:
            return .retry
        }
    }

    /// Whether credentials should be cleared for this error
    var shouldClearCredentials: Bool {
        switch self {
        case .authenticationFailed, .serverError(statusCode: 401, _):
            return true
        default:
            return false
        }
    }

    // MARK: - Equatable

    static func == (lhs: EASError, rhs: EASError) -> Bool {
        switch (lhs, rhs) {
        case (.offline, .offline),
             (.noCredentials, .noCredentials),
             (.authenticationFailed, .authenticationFailed),
             (.accessDenied, .accessDenied),
             (.serverUnavailable, .serverUnavailable),
             (.calendarFolderNotFound, .calendarFolderNotFound),
             (.invalidServerURL, .invalidServerURL),
             (.provisioningDenied, .provisioningDenied),
             (.provisioningRequired, .provisioningRequired):
            return true
        case (.parseError(let a), .parseError(let b)),
             (.networkError(let a), .networkError(let b)),
             (.wbxmlError(let a), .wbxmlError(let b)),
             (.unknown(let a), .unknown(let b)):
            return a == b
        case (.serverError(let codeA, let msgA), .serverError(let codeB, let msgB)):
            return codeA == codeB && msgA == msgB
        default:
            return false
        }
    }
}

/// Suggested recovery action for errors
enum RecoveryAction {
    /// User should retry the operation
    case retry

    /// User should re-authenticate
    case login

    /// User should contact system administrator
    case contactAdmin
}
